goplay/src/main/java/be/tapped/goplay/content/ProgramRepo.kt

```diff
@@ -1,33 +1,100 @@
+            }
+        }
+
+    override suspend fun fetchProgramByLink(link: Program.Link): Either<Failure, Success.Content.Program.Detail> =
+        withContext(Dispatchers.IO) {
+            either {
+                val html = client.safeGet<HttpResponse>("$siteUrl${link.link}").bind().safeReadText().bind()
+                val jsonProgram = programDetailHtmlJsonExtractor.parse(html).bind()
+                val dataObj = catch { jsonSerializer.safeDecodeFromString<JsonObject>(jsonProgram).bind().getValue("data") }.mapLeft(Failure::JsonParsingException).bind()
+                val program = jsonSerializer.safeDecodeFromJsonElement<Program.Detail>(dataObj).bind()
+                Success.Content.Program.Detail(program)
+            }
+        }
+
+    override suspend fun fetchProgramById(id: Program.Id): Either<Failure, Success.Content.Program.Detail> =
+        withContext(Dispatchers.IO) { either { Success.Content.Program.Detail(client.safeGet<Program.Detail>("$siteUrl/api/program/${id.id}").bind()) } }
 
-internal fun httpProgramRepo(client: HttpClient, htmlProgramParser: HtmlProgramParser): ProgramRepo = ProgramRepo {
-    withContext(Dispatchers.IO) {
-        either {
-            val html = client.get<HttpResponse>("$siteUrl/programmas").readText()
-            val programs = htmlProgramParser.parse(html).bind()
-            Success.Content.Programs(programs)
+    override suspend fun fetchPopularPrograms(brand: GoPlayBrand?): Either<Failure, Nel<Success.Content.Program.Detail>> =
+        withContext(Dispatchers.IO) {
+            either {
+                client.safeGet<List<Program.Detail>>(
+                    "$siteUrl/api/programs/popular/" +
+                            when (brand) {
+                                GoPlayBrand.Play4 -> "vier"
+                                GoPlayBrand.Play5 -> "vijf"
+                                GoPlayBrand.Play6 -> "zes"
+                                GoPlayBrand.Play7 -> "zeven"
+                                null -> ""
+                            }
+                )
+                    .bind()
+                    .map(Success.Content.Program::Detail)
+                    .toNel { Failure.Content.NoPrograms }.bind()
+            }
         }
-    }
 }
+
+internal class ProgramDetailHtmlJsonExtractor {
+    private val regex by lazy("data-hero=\"([^\"]+)"::toRegex)
+    internal fun parse(html: String): Either<Failure.HTMLJsonExtractionException, String> =
+        catch {
+            val (htmlEncodedJson) = regex.find(html)!!.destructured
+            htmlEncodedJson.htmlDecode()
+        }.mapLeft(Failure::HTMLJsonExtractionException)
+}
+
+internal class AllProgramsHtmlJsonExtractor {
+    private val regex by lazy("data-program=\"([^\"]+)\""::toRegex)
+    internal fun parse(html: String): Either<Failure, List<String>> =
+        catch(regex.findAll(html).map { it.groupValues[1] }.map(String::htmlDecode)::toList).mapLeft(Failure::HTMLJsonExtractionException)
```

Comment:
lel dat method reference at the end 😁
Nit: I'd use a lambda here

Reply:
👌

Reply:
```suggestion
        catch {
           regex.findAll(html).map { it.groupValues[1] }.map(String::htmlDecode).toList()
         }.mapLeft(Failure::HTMLJsonExtractionException)
```

You're right 😁 I think the `::toList` was a bit overkill, and this reads better IMO.